### PR TITLE
PMM-1081: PS collector doesn't work for queries that don't show every interval

### PR DESCRIPTION
**Pull request by arvenil**
# Issue

Performance Schema Monitor treats queries that don't show up every snapshot as `new queries`.
If query is run in intervals longer than 60s it will be incorrectly scrapped as `new query` and so its full `query count` will be taken as metric but instead the difference to previous value should be taken.

In other words query needs to be executed every interval (e.g. every 60s) to be properly recorded with current implementation.

On pmmdemo we had a query that was run every hour, this clearly doesn't fit 60s interval, so it was incorrectly scrapped as a new query with full counter. 

# Proof

```term
FAIL: perfschema_test.go:659: WorkerTestSuite.Test005

perfschema_test.go:719:
    assert.Equal(t, expect, res)
/home/travis/gopath/src/github.com/percona/qan-agent/vendor/github.com/stretchr/testify/assert/assertions.go:239:
    t.Errorf("%s", "\r"+getWhitespaceString()+labeledOutput(content...))
	Error Trace:	perfschema_test.go:719
	Error:      	Not equal: 
	            	expected: &report.Result{Global:(*event.Class)(0xc420146f60), Class:[]*event.Class{(*event.Class)(0xc420146fc0), (*event.Class)(0xc420147020)}, RateLimit:0x0, RunTime:0, StopOffset:0, Error:""}
	            	actual: &report.Result{Global:(*event.Class)(0xc420146c60), Class:[]*event.Class{(*event.Class)(0xc420146cc0), (*event.Class)(0xc420146d20)}, RateLimit:0x0, RunTime:0, StopOffset:0, Error:""}

	            	Diff:
	            	--- Expected
	            	+++ Actual
	            	@@ -188,3 +188,3 @@
	            	   }),
	            	-  TotalQueries: (uint) 3,
	            	+  TotalQueries: (uint) 1003,
	            	   UniqueQueries: (uint) 2,
	            	@@ -577,3 +577,3 @@
	            	    }),
	            	-   TotalQueries: (uint) 2,
	            	+   TotalQueries: (uint) 1002,
	            	    UniqueQueries: (uint) 0,
```
[source](https://travis-ci.org/percona/qan-agent/jobs/252040366#L674-L697)

# Explanation 

Performance Schema collector works by taking snapshots of `performance_schema.events_statements_summary_by_digest` every `interval` seconds, and then by comparing current snapshot with previous snapshot it calculates the change in metric value.

Let's say snapshots are taken every `60s` (default interval). Below are 3 snapshots and then metrics that are produced from them.  To make data more clear all metrics got removed except `CountStar` which is `query count`. Let's call query with digest `11111111111111111111111111111111` the Query `1` and query with digest `22222222222222222222222222222222` the Query `2`.

* Snapshot 1 - Query `1` was executed `1` time so far, Query `2` was executed `1000` times so far. 
  ```
  [
	{
		"Digest": "11111111111111111111111111111111",
		"CountStar": 1
	},
	{
		"Digest": "22222222222222222222222222222222",
		"CountStar": 1000
	}
  ]
  ```
* Snapshot 2 - Query `2` was not executed during this time frame, so it doesn't show up in the snapshot. Query `1` was executed `1` time in this snapshot so in total `2` times.
  ```
  [
	{
		"Digest": "11111111111111111111111111111111",
		"CountStar": 2
	}
  ]
  ```

* Snapshot 3 - Query `1` was executed `1` time in this snapshot so in total `3` times. Query `2` shows up again and it was executed `2` times in this snapshot so in total `1002` times.
  ```
  [
	{
		"Digest": "11111111111111111111111111111111",
		"CountStar": 3
	},
	{
		"Digest": "22222222222222222222222222222222",
		"CountStar": 1002
	}
  ]
  ```

So we have `CountStar` (`query count`):

* First interval:
  * Query `1`: `1`
  * Query `2`: `1000`
* Second interval:
  * Query `1`: `2`
* Third interval:
  * Query `1`: `3`
  * Query `2`: `1002`

Now to calculate metrics we need to have 2 snapshots, so we skip first interval. We can create metrics only between:
* 1st and 2nd snapshot
  * Query `1` SUM: `2 - 1 = 1`
  * Query `2` SUM: metric is not generate for this because we don't have it in second snapshot, so we have nothing to compare too.
* 2nd and 3rd snapshot
  * Query `1` SUM: `3 - 2 = 1`
  * **Query `2` SUM: since Query `2` doesn't exist in previous snapshot then it's thought to be new query that was executed `1002` times during this interval. This is wrong as if we compare it to first snapshot we then know it was executed `2` times not `1002`.**

So in summary:
* metric1 - skipped, can't be generated from one snapshot
* metric2:
  * Query `1` SUM: `1`
* metric3:
  * Query `1` SUM: `1`
  * **Query `1` SUM: `1002` but it should be `2`**

# Fix

Replicate `events_statements_summary_by_digest` table.

On `performance schema collector`/`qan-agent` start we could scrap whole `events_statements_summary_by_digest` table and store it in-memory or locally, and when comparing new snapshots to it we should update it at the end with new values.

```
take `base snapshot` of whole `events_statements_summary_by_digest` table
for {
  take `interval snapshot`
  compare `interval snapshot` to `base snapshot` and generate difference as metrics
  update `base snapshot` (e.g. with values from `interval snapshot`) \
      to sync it with `events_statements_summary_by_digest` table
}
```

# Results

5 days chart shows significant drop in QPS after fix was applied
percona_query_analytics

1 day chart shows 140 QPS and we can see hourly spikes have similar values through whole day.
percona_query_analytics

1 hour chart shows 165 QPS and hourly spike is clearly visible.
percona_query_analytics

### CODE REVIEW

**Comment by arvenil** [comment]
@AlekSi If you have time and wish, could you review this PR? This is one of the most separate part of the agent and it's not so complex. I think it's a good starting point to understand some of the agent parts and it would be nice to get third point of view for qan-agent.

**Comment by AlekSi** [comment]
OMG, why it uses both go-check and testify?

File: qan/analyzer/mysql/worker/perfschema/worker.go [hunk 1]
```diff
@@ -455,79 +527,67 @@ CLASS_LOOP:
 		// Each row is an instance of the query executed in the schema.
 	ROW_LOOP:
```
Comment (AlekSi) [inline qan/analyzer/mysql/worker/perfschema/worker.go]: `RowLoop`

**Comment by arvenil** [comment]
@AlekSi gocheck dropped

File: qan/analyzer/mysql/worker/perfschema/perfschema_test.go [hunk 2]
```diff
@@ -156,186 +185,334 @@ func normalizeResult(res *report.Result) {
...
 
-func (s *WorkerTestSuite) Test001(t *C) {
+func test001(t *testing.T, logger *pct.Logger, nullmysql *mock.NullMySQL) {
 	// This is the simplest input possible: 1 query in iter 1 and 2. The result
 	// is just the increase in its values.
 
-	rows, err := s.loadData("001")
-	t.Assert(err, IsNil)
+	rows, err := loadData("001")
+	require.Nil(t, err)
```
Comment (AlekSi) [inline qan/analyzer/mysql/worker/perfschema/perfschema_test.go]: `require.NoError(t, err)`

**Review by rnovikovP — APPROVED** [review]

**Pull request merged**